core/shared/src/main/scala/fs2/interop/flow/StreamProcessor.scala

```diff
@@ -0,0 +1,44 @@
+package fs2
+package interop
+package flow
+
+import java.util.concurrent.Flow
+import cats.effect.{Async, Resource}
+
+private[flow] final class StreamProcessor[F[_], I, O](
+    streamSubscriber: StreamSubscriber[F, I],
+    streamPublisher: StreamPublisher[F, O]
+) extends Flow.Processor[I, O] {
+  override def onSubscribe(subscription: Flow.Subscription): Unit =
+    streamSubscriber.onSubscribe(subscription)
+
+  override def onNext(i: I): Unit =
+    streamSubscriber.onNext(i)
+
+  override def onError(ex: Throwable): Unit =
+    streamSubscriber.onError(ex)
+
+  override def onComplete(): Unit =
+    streamSubscriber.onComplete()
+
+  override def subscribe(subscriber: Flow.Subscriber[? >: O <: Object]): Unit =
+    streamPublisher.subscribe(subscriber)
+}
+
+private[flow] object StreamProcessor {
+  def fromPipe[F[_], I, O](
+      pipe: Pipe[F, I, O],
+      chunkSize: Int
+  )(implicit
+      F: Async[F]
+  ): Resource[F, StreamProcessor[F, I, O]] =
+    for {
+      streamSubscriber <- Resource.eval(StreamSubscriber[F, I](chunkSize))
+      inputStream = streamSubscriber.stream(subscribe = F.unit)
```

Comment:
We don't have access to the upstream `Publisher` yet, so we can't do the `subscribe` here. So we do nothing and wait for it.

Reply:
> So we do nothing and wait for it.

I'm not entirely sure I understand the waiting part. Can/should we use a `Deferred` here?

Reply:
No sorry, my point is that `subscribe` should be something like `IO(publisher.subscribe(streamSubscriber))` as you can see in the implementation of `syntax.fromPublisher`.
Here, however, we don't have a `Publisher` yet, so we don't control when we are published.

Basically, this is very similar to the `fromPublisher` overload that accepts a `Subscriber => F[Unit]`, just that rather than receiving the lambda, we are just returning a raw `Processor / Subscriber`.

Reply:
Thanks, I understand now.

I think what confused me was that this is the first time we are effectively exposing `StreamSubscriber`. Our existing `{to,from}Publisher` APIs have kept it hidden.

So I do feel weird about exposing it now. But I guess nothing here is unreasonable 🤔